doc/api/next_api_changes/behavior/23573-AL.rst

```diff
@@ -0,0 +1,6 @@
+All Axes have a ``get_subplotspec`` method now, which returns None for Axes not positioned via a gridspec
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Previously, this method was only present for Axes positioned via a gridspec.
+Following this change, checking ``hasattr(ax, "get_gridspec")`` or
+``isinstance(ax, SubplotBase)`` should now be replaced by
```

Comment:
I'm +1 of getting rid of the separate Subplot classes.

However, if I understand this correctly, both ``hasattr(ax, "get_gridspec")`` or
``isinstance(ax, SubplotBase)`` change behavior under this PR in that the are True also for non-suplots. This is an API break, we cannot afford without deprecation. (There are ~600 matches / ~1700 matches on github (some in Matplotlib, but also in other libs).

We may also consider whether the subplot info needs to be stored / be accessible at all from the Axes. I'm on holidays, so not able to check code details, but it may be feasible to hold the arrangement info outside of the Axes. The arrangement is basically a figure-level concept. We could make it accessible from there and query it for information regarding a specific Axes, i.e. something like `fig.get_gridspec().get_subplotspec(ax)`.

Reply:
Fixing `isinstance(ax, SubplotBase)` to make it work is relatively easy (via `__instancecheck__`), and now done.
I can certainly also fix `hasattr(ax, "get_gridspec")` to make that return False for absolute-positioned axes, but that basically depends on what you think the end plan should be.  Do you want to keep the hasattr check "for ever"?  If not, I can't see a proper deprecation path: certainly, tricks based on `__getattribute__` or `@property` can let me raise a DeprecationWarning whenever one does `hasattr(ax, "get_gridspec")`, but what is the alternative for them?  They can't get rid of the warning by doing `ax.get_gridspec() is None` instead, because this will always also raise the DeprecationWarning due to how hasattr works.  Keeping the hasattr check working forever seems slightly inelegant though (it won't be a normal method, for example).
`fig.get_gridspec().get_subplotspec(ax)` probably can't be made to work (as you can have multiple gridspecs per figure) but `fig.get_subplotspec(ax)` probably can (and then we can just wholly deprecate `ax.get_gridspec`/`ax.get_subplotspec` in all cases), but that's a bigger refactor and probably @jklymak should comment on that too.

Reply:
I agree the breakage will typically be loud anyways (i.e. even if the hasattr check incorrectly succeeds, a later stage will likely fail).
I don't think an "axes out there that does not use our base class" is really likely to exist at all -- there are far too many private attributes that need to be set up correctly for that to be realistic...